From: methanal <baclofen@tuta.io>
Date: Wed, 14 Jun 2023 03:15:27 +0900
Subject: Input: synaptics-rmi4 - handle zero electrode count

Some replacement displays include third-party touch ICs which report
the electrode count as 0. Reading a length of zero bytes via i2c
confuses the bus and results in a crash. Introduce checks for the
electrode count, and skip the reads if they are zero.

Signed-off-by: methanal <baclofen@tuta.io>
---
 drivers/input/rmi4/rmi_f55.c | 44 ++++++++++++++++++++++++--------------------
 1 file changed, 24 insertions(+), 20 deletions(-)

diff --git a/drivers/input/rmi4/rmi_f55.c b/drivers/input/rmi4/rmi_f55.c
index 488adac..8da8a0a 100644
--- a/drivers/input/rmi4/rmi_f55.c
+++ b/drivers/input/rmi4/rmi_f55.c
@@ -70,30 +70,34 @@ static int rmi_f55_detect(struct rmi_function *fn)
 		 * enabled electrodes is the sum of all field entries with a
 		 * value other than 0xff.
 		 */
-		error = rmi_read_block(fn->rmi_dev,
-				       fn->fd.control_base_addr + 1,
-				       buf, f55->num_rx_electrodes);
-		if (!error) {
-			total = 0;
-			for (i = 0; i < f55->num_rx_electrodes; i++) {
-				if (buf[i] != 0xff)
-					total++;
+		if (f55->num_rx_electrodes) {
+			error = rmi_read_block(fn->rmi_dev,
+					       fn->fd.control_base_addr + 1,
+					       buf, f55->num_rx_electrodes);
+			if (!error) {
+				total = 0;
+				for (i = 0; i < f55->num_rx_electrodes; i++) {
+					if (buf[i] != 0xff)
+						total++;
+				}
+				f55->cfg_num_rx_electrodes = total;
+				drv_data->num_rx_electrodes = total;
 			}
-			f55->cfg_num_rx_electrodes = total;
-			drv_data->num_rx_electrodes = total;
 		}
 
-		error = rmi_read_block(fn->rmi_dev,
-				       fn->fd.control_base_addr + 2,
-				       buf, f55->num_tx_electrodes);
-		if (!error) {
-			total = 0;
-			for (i = 0; i < f55->num_tx_electrodes; i++) {
-				if (buf[i] != 0xff)
-					total++;
+		if (f55->num_tx_electrodes) {
+			error = rmi_read_block(fn->rmi_dev,
+					       fn->fd.control_base_addr + 2,
+					       buf, f55->num_tx_electrodes);
+			if (!error) {
+				total = 0;
+				for (i = 0; i < f55->num_tx_electrodes; i++) {
+					if (buf[i] != 0xff)
+						total++;
+				}
+				f55->cfg_num_tx_electrodes = total;
+				drv_data->num_tx_electrodes = total;
 			}
-			f55->cfg_num_tx_electrodes = total;
-			drv_data->num_tx_electrodes = total;
 		}
 	}
 
